### DISCUSSION TITLE
generateStaticSearchParams()

### DISCUSSION THREAD

**Discussion by stefan-garofalo** [vercel/next.js#58181] [App Router]
Currently we have `generateStaticParams()` to create dynamic segments and generate pages at build time.
Often we want to store state in the url and I find that in many use cases we only need a few params with limited values.

For example I am building a page displaying articles, and I want to query them by date. Therefore I will only have a `date` search param with the following values: `2020, 2021, 2022, 2023`.
Since I am deciding ahead of time what values there could possibly be then it should be possible to:

```javascript
export async function generateStaticSearchParams() {
  return [{ date: '2020' }], { date: '2021' }, { date: '2022' }, { date: '2023' }]
}
```

Since currently this behavior is not available, could it be implemented? Is there a reason it might not be possible?

**Comment by icyJoseph** [comment]
I think you can do this with rewrites... and by not allowing direct access to the "destination" file:

- https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites#header-cookie-and-query-matching

With the above, you can match requests with a given query param, and rewrite them to a destination, using the query param, as long as you capture it, as they mention in that document.

Then you'd statically generate `/path/[date]` with the array you show, and use that as the destination for the rewrites.

**Reply by stefan-garofalo** [reply]
Probably it is a good enough workaround , but Id like more a proper api

**Reply by stefan-garofalo** [reply]
> I think you can do this with rewrites... and by not allowing direct access to the "destination" file:
> 
> 
> 
> - https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites#header-cookie-and-query-matching
> 
> 
> 
> With the above, you can match requests with a given query param, and rewrite them to a destination, using the query param, as long as you capture it, as they mention in that document.
> 
> 
> 
> Then you'd statically generate `/path/[date]` with the array you show, and use that as the destination for the rewrites.
> 
> 
> 
> 

Actually if you have more than one search parameter you end up with multiple nested segments, it doesn't scale. The tweak is too specific

**Comment by amannn** [comment]
+1 for this. The more general issue for me is not being able to have caching headers on any page that reads `searchParams`.

**Comment by AndonMitev** [comment]
Yeah having a option to generateStaticSearchParams would give even more life to the ISR
